### PR TITLE
#5949 - Update Editor: Preview Markdown

### PR DESCRIPTION
**Pull request by Miaplacidus**
This PR creates a markdown preview tab (as opposed to the existing markdown preview button present in the toolbar), enforces that all threads/comments going forward will be in markdown, and provides minimal formatting for when a user wants to edit a rich text post (i.e. adds tags for headers, italic text, bold text, and strikethroughs).

Note that the formatting is not intended to provide a completely seamless transition from rich text to markdown, i.e. some formatting will be lost. The intent is to provide users with a place to start so they can make any necessary edits with greater ease. In addition, this PR does not include ALL of the functionality mentioned in the designs. That is left to other cards. Look out only for the features mentioned in the previous paragraph. 

## Link to Issue
Closes: #5949 

## Description of Changes
- replaces markdown preview modal with preview tab
- provides basic formatting when editing rich text threads and comments

## Test Plan
- attempt to edit a rich text thread. You should see some formatting added (#, *, **, ~~) to the text in the editor
- you should be able to view a preview of markdown text under the markdown tab

### CODE REVIEW

**Review by masvelio — COMMENTED** [review]
1. open a thread in community that you are not part of. The editor is in disabled state, however:
- it has wrong tooltip text - it should say "join community to comment"
- despite it is blocked, there is an initial focus so I am able to type text (prevent focus if disabled)
- I am able to submit comment but I am getting error anyways (will be fixed by the point from above 👆 )

https://github.com/hicommonwealth/commonwealth/assets/14819225/7f8062c2-301c-4597-a1f6-c50c3e111bee

2. When I want to edit rich text thread
- warning modal is rendered 2 times (i need to click twice to get rid of it)
- despite I am not accepting it, the editor still is in edit mode. If I click X, ESC or NO, modal should disappear and the editor should not be in edit mode 

https://github.com/hicommonwealth/commonwealth/assets/14819225/c2d65b11-e0a6-4bfd-9597-1de38def1cd1

3. Not sure if this is part of this PR but options in markdown does not seem to work well. When I use bold, italic, strikethrough etc, the cursor should be between the special characters, not after. Number, order and todo list does not work for me at all, unless I select the text first

https://github.com/hicommonwealth/commonwealth/assets/14819225/5c4f6f95-7c29-4e1c-87b6-138ea1ae0be7

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_component_showcase.tsx [hunk 1]
```diff
@@ -1213,6 +1224,21 @@ export const ComponentShowcase = () => {
             />
           ))}
         </CWTabsRow>
+
+        <CWText type="h3">Boxed Tabs</CWText>
+        <CWTabsRow boxed={true}>
```
Comment (lzach83) [inline packages/commonwealth/client/scripts/views/components/component_kit/cw_component_showcase.tsx:1229]: `boxed={true}` is redundant just the presence of `boxed` as a prop renders it truthy

File: packages/commonwealth/client/scripts/views/components/react_quill_editor/toolbar.tsx [hunk 2]
```diff
@@ -51,22 +49,15 @@ const LIST_ITEM_PREFIX = {
...
 export const CustomQuillToolbar = ({
   toolbarId,
-  setIsPreviewVisible,
-  handleToggleMarkdown,
-  isMarkdownEnabled,
   isDisabled = false,
-  isPreviewDisabled,
 }: CustomQuillToolbarProps) => (
   <div id={toolbarId} className="CustomQuillToolbar">
+    <div className="tabs"></div>
```
Comment (lzach83) [inline packages/commonwealth/client/scripts/views/components/react_quill_editor/toolbar.tsx:60]: Is this correct?

**Review by lzach83 — DISMISSED** [review]

**Comment by Israellund** [comment]
I agree with everything Marcin and Zach commented. Once these are fixed I'm happy to approve.

File: packages/commonwealth/client/scripts/views/components/react_quill_editor/react_quill_editor.tsx [hunk 3]
```diff
@@ -1,15 +1,14 @@
...
 import { RangeStatic } from 'quill';
 import ImageUploader from 'quill-image-uploader';
 import MagicUrl from 'quill-magic-url';
 import React, { useEffect, useMemo, useRef, useState } from 'react';
 import { DragDropContext, Droppable } from 'react-beautiful-dnd';
 import ReactQuill, { Quill } from 'react-quill';
-
-import { nextTick } from 'process';
 import { openConfirmation } from '../../modals/confirmation_modal';
-import { PreviewModal } from '../../modals/preview_modal';
```
[the remark this answers is not in the document]
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/react_quill_editor/react_quill_editor.tsx:11]: Good catch! Deleted!

**Comment by Miaplacidus** [comment]
@masvelio @lzach83 @Israellund @ForestMars 
The cursor being between the formatting characters has been moved to issue #6598 as a) directly manipulating the Quill cursor isn't actually possible b) the workarounds are difficult and will only theoretically work, i.e. I'm not sure even that ticket is possible.

**Comment by Miaplacidus** [comment]
@masvelio @Israellund @lzach83 @CowMuon 
The issues mentioned in the comment have been addressed. 

I believe the "doubled" appearance of the rich text to markdown warning modal has to do with the way we handle rendering in local development (based on previous experience with `useEffect` running twice locally).

**Review by jnaviask — DISMISSED** [review]
Noticed this bug that seems to relate to the double modal: https://www.loom.com/share/8cd3c680906c4c8b83f1701d9198bc76

We should figure this one out.

**Comment by lzach83** [comment]
> Noticed this bug that seems to relate to the double modal: https://www.loom.com/share/8cd3c680906c4c8b83f1701d9198bc76
> 
> We should figure this one out.

Currently working through this now.

**Comment by lzach83** [comment]
Converting to draft for now, theres a few bugs I need to sort through before this can be reviewed again.

**Review by gdjohnson — APPROVED** [review]
The areas Marcin flagged all seem to work for me—the Conversion and Cancel modals are all singletons; the editor is properly disabled for out-of-towners; and a rough RichText to Markdown conversion works (more or less) on edit. 

I do think that the border box issue and the expanding/contracting editor/preview transition should have tickets made for them—there's a related off-by-one-pixel problem with how the in-focus border renders. But that's a CSS issue, not a problem with core functionality.

**Review by Israellund — APPROVED** [review]
LGTM

**Pull request merged**